program/rust/src/processor/upd_price.rs

```diff
@@ -161,37 +162,76 @@ pub fn upd_price(
+
+        {
+            let mut price_data =
+                load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+            let publisher_price = &mut price_data.comp_[publisher_index].latest_;
+            publisher_price.price_ = cmd_args.price;
+            publisher_price.conf_ = cmd_args.confidence;
+            publisher_price.status_ = status;
+            publisher_price.pub_slot_ = cmd_args.publishing_slot;
+        }
+    }
+
+    // get number of slots from last update
+    let slots_since_last_update = clock.slot - latest_aggregate_price.pub_slot_;
+
+    // If the price update is the first in the slot and the aggregate is trading, update the previous slot, price, conf, and timestamp.
+    if slots_since_last_update > 0 && latest_aggregate_price.status_ == PC_STATUS_TRADING {
+        let mut price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+        price_data.prev_slot_ = price_data.agg_.pub_slot_;
+        price_data.prev_price_ = price_data.agg_.price_;
+        price_data.prev_conf_ = price_data.agg_.conf_;
+        price_data.prev_timestamp_ = clock.unix_timestamp;
+    };
+
+
+    let updated = unsafe {
+        // NOTE: c_upd_aggregate must use a raw pointer to price
+        // data. Solana's `<account>.borrow_*` methods require exclusive
+        // access, i.e. no other borrow can exist for the account.
+        c_upd_aggregate(
+            price_account.try_borrow_mut_data()?.as_mut_ptr(),
+            clock.slot,
+            clock.unix_timestamp,
+        )
+    };
+
+    // If the aggregate was successfully updated, calculate the difference and update TWAP.
+    if updated {
+        let agg_diff = (clock.slot as i64)
+            - load_checked::<PriceAccount>(price_account, cmd_args.header.version)?.prev_slot_
+                as i64;
+        {
             let mut price_data =
                 load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+            // check if its the first price update in the slot
+            if slots_since_last_update > 0 {
+                price_data.prev_twap_ = price_data.twap_;
+                price_data.prev_twac_ = price_data.twac_;
+                price_data.prev_price_cumulative = price_data.price_cumulative;
+            }
+            price_data.twap_ = price_data.prev_twap_;
+            price_data.twac_ = price_data.prev_twac_;
+            price_data.price_cumulative = price_data.prev_price_cumulative;
+            price_data.update_price_cumulative()?;
             // We want to send a message every time the aggregate price updates. However, during the migration,
             // not every publisher will necessarily provide the accumulator accounts. The message_sent_ flag
             // ensures that after every aggregate update, the next publisher who provides the accumulator accounts
             // will send the message.
             price_data.message_sent_ = 0;
```

Comment:
(maybe future PR) I'm pretty sure we don't need `message_sent_` anymore, since now everytime a publisher publishes we're gonna attempt an aggregation and attempt to send